Sources/Basic/Path.swift

```diff
@@ -14,128 +14,59 @@ import POSIX
-                chars.replaceSubrange(chars.startIndex ..< nEnd, with: homeDir)
-            }
-        }
-        
-        // At this point we expect to have a path separator as first character.
-        assert(chars.first == PathSeparatorCharacter)
-
-        // Split the character array into parts, folding components as we go.
-        // As we do so, we count the number of characters we'll end up with in
-        // the normalized string representation.
-        var parts: [String.CharacterView] = []
-        var capacity = 0
-        for part in chars.split(separator: PathSeparatorCharacter) {
-            switch part.count {
-              case 0:
-                // Ignore empty path components.
-                continue
-              case 1 where part.first == ".":
-                // Ignore `.` path components.
-                continue
-              case 2 where part.first == "." && part.last == ".":
-                // If there's a previous part, drop it; otherwise, do nothing.
-                if let prev = parts.last {
-                    parts.removeLast()
-                    capacity -= prev.count
-                }
-              default:
-                // Any other component gets appended.
-                parts.append(part)
-                capacity += part.count
-            }
-        }
-        capacity += max(parts.count, 1)
-        
-        // Create an output buffer using the capacity we've calculated.
-        // FIXME: Determine whether this is the most efficient way to do it.
-        var result = ""
-        result.reserveCapacity(capacity)
-        
-        // Put the normalized parts back together again.
-        var iter = parts.makeIterator()
-        result.append(PathSeparatorCharacter)
-        if let first = iter.next() {
-            result.append(contentsOf: first)
-            while let next = iter.next() {
-                result.append(PathSeparatorCharacter)
-                result.append(contentsOf: next)
-            }
-        }
-        
-        // Sanity-check the result (including the capacity we reserved).
-        assert(!result.isEmpty, "unexpected empty string")
-        assert(result.characters.count == capacity, "count: " +
-            "\(result.characters.count), cap: \(capacity)")
-        
-        // Use the result as our stored string.
-        _impl = PathImpl(string: String(result))
+        // Normalize the absolute string and store it as our PathImpl.
+        _impl = PathImpl(string: normalize(absolute: string))
     }
     
+    /// Initializes the AbsolutePath from an AbsolutePath and a RelativePath.
+    public init(_ absPath: AbsolutePath, _ relPath: RelativePath) {
+        // Both paths are already normalized, so we just construct a new path.
+        let absStr = absPath._impl.string
+        let relStr = relPath._impl.string
+        _impl = PathImpl(string: relStr == "." ? absStr :
+            absStr + String(pathSeparatorCharacter) + relStr)
```

Comment:
This needs to also handle the case where the relative path starts with "..", I think?